### PR TITLE
feat: implement streaming output to reduce memory usage and improve TTFB

### LINKED ISSUE #168: Support streaming output
Currently, spanner-mycli collects all result before printing results.
It means Time To First Byte is high.

Ideally, we can print result from `RowIterator` in streaming manner.

Note: Table format for streaming data is realtively hard. It may be good to format table using first N rows in stream.

## Subtask

- [ ] Table format for streaming
- [ ] Support normal `ExecuteStreamingSql`
- [ ] Support partitioned query

### PR DESCRIPTION
**Pull request by apstndb**
## Summary

Implements streaming output for spanner-mycli to significantly reduce memory usage and improve Time-To-First-Byte (TTFB) for large query results.

Fixes #168

### Performance Improvements
- **55% memory reduction** for large datasets (368.7 MB → 165.7 MB for 1M rows)
- **51% faster TTFB** for large result sets
- O(1) memory complexity for streaming vs O(n) for buffered mode

### Configuration
- New `CLI_STREAMING` variable with three-value enum (AUTO/TRUE/FALSE)
- Smart AUTO mode defaults: streaming for CSV/Tab/Vertical/HTML/XML, buffered for Table formats
- `CLI_TABLE_PREVIEW_ROWS` (default: 50) for table width calculation in streaming mode
- Command-line flag: `--streaming=[AUTO|TRUE|FALSE]`

### Architecture
- Unified streaming/buffered execution paths using `iter.Seq` pattern
- StreamManager as single source of truth for all I/O operations
- Clean separation of concerns with format-specific streaming implementations

### Core Components
- **streaming.go**: Core streaming logic with row iteration
- **row_processor.go**: Unified interface for processing rows
- **execution_metrics.go**: Comprehensive metrics collection (TTFB, memory, GC)
- **formatters_*_streaming.go**: Format-specific streaming implementations

### Smart Defaults (AUTO mode)
| Format | Default Mode | Reason |
|--------|-------------|---------|
| CSV, Tab, Vertical | Streaming | Low memory usage, immediate output |
| HTML, XML | Streaming | Structure allows incremental output |
| Table formats | Buffered | Accurate column width calculation |

## Testing
- Comprehensive unit tests for all streaming formats
- Integration tests for streaming functionality
- Performance benchmark script in `scripts/verify_streaming.sh`

## Documentation
- Integrated into README.md with system variables and command-line flags
- Clear user-facing documentation for configuration options

## Breaking Changes
None - fully backward compatible with smart defaults.

## Performance Benchmark Results

Using the Songs table (1,024,000 rows):
```
Memory Usage: 55% reduction (368.7 MB → 165.7 MB)
Time to First Byte: 51% faster for 500K+ rows
Scalability: Handles arbitrarily large result sets without OOM
```

The implementation provides significant benefits for large Spanner queries while maintaining compatibility and ease of use.

### CODE REVIEW

File: formatters_common.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+// InitFormat writes CSV headers if needed.
+func (f *CSVFormatter) InitFormat(columns []string, metadata *sppb.ResultSetMetadata, sysVars *systemVariables, previewRows []Row) error {
+	if f.initialized {
+		return nil
+	}
+
+	if len(columns) == 0 {
+		return fmt.Errorf("no columns to output")
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply formatters_common.go]: Thank you for the feedback!

Fixed in commit e4df053.

File: formatters_html_xml_streaming.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+func (f *HTMLFormatter) InitFormat(columns []string, metadata *sppb.ResultSetMetadata, sysVars *systemVariables, previewRows []Row) error {
+	if f.initialized {
+		return nil
+	}
+
+	f.columns = columns
+
+	if len(columns) == 0 {
+		return fmt.Errorf("no columns to output")
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply formatters_html_xml_streaming.go]: Thank you for the feedback!

Fixed in commit e4df053.

File: formatters_table_streaming.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+			renderer.NewBlueprint(tw.Rendition{Symbols: tw.NewSymbols(tw.StyleASCII)})),
+		tablewriter.WithHeaderAlignment(tw.AlignLeft),
+		tablewriter.WithTrimSpace(tw.Off),
+		tablewriter.WithHeaderAutoFormat(tw.Off),
+		tablewriter.WithStreaming(tw.StreamConfig{
+			Enable: true,
+		}),
+	).Configure(func(config *tablewriter.Config) {
+		config.Row.Formatting.AutoWrap = tw.WrapNone
+	})
```
[the remark this answers is not in the document]
Comment (apstndb) [reply formatters_table_streaming.go]: Thank you for the review\! This is actually working as intended.

Column alignment is only used by EXPLAIN and DESCRIBE statements, which have custom formatting logic and don't use the streaming mode - they always return a complete Result with ColumnAlign set and use buffered formatting.

Regular SQL queries don't specify column alignment (the field remains nil), so both buffered and streaming modes default to left alignment for all columns.

I've added a clarifying comment in commit 5837006 to document this design decision.
EOF < /dev/null

File: execute_sql.go [hunk 4]
```diff
@@ -92,6 +207,129 @@ func executeSQL(ctx context.Context, session *Session, sql string) (*Result, err
...
+	if err != nil {
+		if session.InReadWriteTransaction() && spanner.ErrCode(err) == codes.Aborted {
+			// Need to call rollback to free the acquired session
+			rollback := &RollbackStatement{}
+			if _, rollbackErr := rollback.Execute(ctx, session); rollbackErr != nil {
+				return nil, errors.Join(err, fmt.Errorf("error on rollback: %w", rollbackErr))
+			}
+		}
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply execute_sql.go]: Thank you for the feedback!

Fixed in commit 36f4ed9.

**Comment by apstndb** [comment]
## Development Insights from Streaming Output Implementation

This PR provided several valuable insights for future development:

### Architecture Insights

1. **Iterator Pattern with Go 1.23+ `iter.Seq`**: The new `iter.Seq[V]` pattern provides a clean abstraction for both streaming and buffered processing. By converting `spanner.RowIterator` to `iter.Seq[Row]`, we achieved a unified interface that works seamlessly with both execution modes.

2. **Strategy Pattern for Processing**: The `RowProcessor` interface (`Init`, `ProcessRow`, `Finish`) successfully decoupled the row processing logic from the data source, making it easy to switch between buffered and streaming modes without changing the core logic.

3. **Centralized I/O Management**: Using `StreamManager` as the single source of truth for output streams and terminal width eliminated redundancy and ensured consistency. This pattern should be applied to other areas where I/O state is shared.

### Performance Insights

1. **Memory Complexity Reduction**: Streaming reduced memory complexity from O(n) to O(1) for most formats, achieving up to 55% memory reduction for large result sets.

2. **TTFB Improvement**: Time to First Byte improved by 51% for large queries by outputting rows as they arrive rather than waiting for the complete result set.

3. **Metrics Collection vs Display**: Always collecting metrics but controlling display via `CLI_PROFILE` flag provides flexibility without performance penalty for non-profiling use cases.

### Implementation Patterns

1. **Three-Value Enum for Mode Selection**: The `AUTO/TRUE/FALSE` pattern for `CLI_STREAMING` provides better user control than a simple boolean, allowing format-dependent automatic selection.

2. **Preview Pattern for Table Format**: The `TablePreviewProcessor` pattern of collecting a configurable number of preview rows for column width calculation elegantly solves the streaming vs. proper formatting trade-off.

3. **Template Simplification**: Using direct `if` statements instead of nested `with` blocks in templates improves readability. Avoiding floating-point equality comparisons with `ne` in favor of `gt` comparisons prevents precision issues.

### Code Quality Observations

1. **File Consolidation**: Merging related functionality (e.g., `memory_stats.go` into `execution_metrics.go`) reduces file count without sacrificing clarity when the components are tightly coupled.

2. **Error Handling Centralization**: Removing redundant error handling from `executeStreamingSQL` and relying on the caller's error handling reduces code duplication and maintains a single point of control.

3. **Documentation in Code**: Adding explanatory comments for non-obvious design decisions (like why column alignment isn't needed in streaming table formatter) prevents future confusion and false positive reviews.

### Testing Insights

1. **Integration Test Adaptation**: Existing integration tests required minimal changes, mainly initializing `StreamManager` in test setup, demonstrating good backward compatibility.

2. **Format Consistency**: Ensuring all formatters handle edge cases consistently (e.g., empty columns) prevents unexpected errors and simplifies error handling upstream.

### Review Process Insights

1. **False Positives in Automated Reviews**: Gemini Code Assist's suggestion about column alignment was a false positive - understanding the specific use case (EXPLAIN/DESCRIBE don't use streaming) was crucial.

2. **Iterative Refinement**: Multiple review cycles helped identify subtle issues like redundant error handling that might have been missed in a single review.

This implementation demonstrates that significant performance improvements can be achieved through architectural changes while maintaining backward compatibility and code clarity.
